packages/api-request-builder/src/default-params.js

```diff
@@ -62,23 +63,104 @@ export function getDefaultSearchParams (): ServiceBuilderDefaultParams {
 export function setDefaultParams () {
   this.params.expand = getDefaultQueryParams().expand
 
-  if (this.features.indexOf(features.queryOne) >= 0)
+  if (this.features.includes(features.queryOne))
     this.params.id = getDefaultQueryParams().id
 
-  if (this.features.indexOf(features.query) >= 0) {
+  if (this.features.includes(features.query)) {
     this.params.pagination = getDefaultQueryParams().pagination
     this.params.query = getDefaultQueryParams().query
   }
 
-  if (this.features.indexOf(features.search) >= 0) {
+  if (this.features.includes(features.search)) {
     this.params.staged = getDefaultSearchParams().staged
     this.params.pagination = getDefaultSearchParams().pagination
     this.params.search = getDefaultSearchParams().search
   }
 
-  if (this.features.indexOf(features.projection) >= 0)
+  if (this.features.includes(features.projection))
     this.params.staged = true
 
-  if (this.features.indexOf(features.suggest) >= 0)
+  if (this.features.includes(features.suggest))
     this.params.searchKeywords = []
 }
+
+const hasKey = (obj: {}, key: string): boolean =>
+  Object.prototype.hasOwnProperty.call(obj, key)
+
+/**
+ * Set the supplied parameters given the current service object.
+ *
+ * @return {void}
+ */
+export function setParams (params: ServiceBuilderParams) {
```

Comment:
Be careful ⚠️

Depending on how the service was configured with the "features" you might have those internal methods or not (`this.xxx`).
This means that not all params can be parsed. I don't think it's a problem, because if you use the service for e.g. products you won't use e.g. `byCustomerId` so it does not make sense to pass it and parse it.

Nevertheless, we need to check that the setter function is defined before calling it

Reply:
I'm aware of that and deliberately set it up like that.

As I pointed out in https://github.com/commercetools/nodejs/pull/337#discussion_r149602263 the `parse` is the declarative counterpart to the imperative API. So they should behave the same.

When using the imperative API and trying to use `byCustomerId` on a service that does not support it, the user is calling a non-existent `byCustomerId` function and an error will be thrown: `service.byCustomerId(2).build()`.

At the moment the same thing will happen when the user uses the declarative API: `service.parse({ customerId: 2 }).build()`.

So with this implementation they already have the same behaviour. So I would not change anything about it.

> Nevertheless, we need to check that the setter function is defined before calling it

That would be like adding `noop` alternatives to the imperative API in case the a feature is not enabled. So it would fail silently, which is pretty bad. The same applies for the declarative API, which is why I would not add these checks.

Reply:
I guess this relates to the other "throwing" part. Makes sense yes, thanks for the explanation 👍